File: AUGUST-2025/08-08-2025/p808.swift
class Solution {
    func soupServings(_ n: Int) -> Double {

        guard n < 5000 else { return 1 }

        var cache = [Int: [Int: Double]]()
        
        func sim(_ a: Int = n, _ b: Int = n) -> Double {
            let a = max(a, 0)
            let b = max(b, 0)

            if let c = cache[a]?[b] { return c }
            let res: Double
            
            switch (a, b) {
            case (0, 0): res = 0.5
            case (0, _): res = 1
            case (_, 0): res = 0
            default: res = (
                sim(a - 100, b) +
                sim(a - 75, b - 25) +
                sim(a - 50, b - 50) +
                sim(a - 25, b - 75)
            ) / 4
            }

            cache[a, default: [:]][b] = res
            return res
        }

        return sim()
    }
}

File: AUGUST-2025/08-08-2025/p808.rb
class Numeric
    def max(v) = self < v ? v : self
    def min(v) = self > v ? v : self
end

OPS = [[100, 0], [75, 25], [50, 50], [25, 75]]

def soup_servings(n)
    return 0.5 if n.zero?
    return 1.0 if n >= 4800
    prob, cur = 0.0, {n => {n => 1.0}}
    while cur.any? {|k, v| k.positive? || v.any? {|vk, vv| vk.positive?} } do
        cur = cur.each_with_object(Hash.new {|h, k| h.update(k => Hash.new(0.0))}) {|(a, v), res|
            v.each {|b, p|
                OPS.each {|(da, db)|
                    newa, newb, newp = 0.max(a - da), 0.max(b - db), p / 4.0
                    prob += newp / (newb.zero? ? 2 : 1) if a.positive? && newa.zero? && !b.zero?
                    res[newa][newb] # for proper initialization
                    res[newa][newb] += newp if [newa, newb].all?(&:positive?)
                }
            }
        }
    end
    prob
end